src/System.Private.ServiceModel/src/Internals/System/Runtime/AsyncLock.cs

```diff
@@ -4,139 +4,146 @@
 
 using System.Threading;
 using System.Threading.Tasks;
-#if DEBUG
-using System.Diagnostics;
-#endif
+using Microsoft.Extensions.ObjectPool;
 
 namespace System.Runtime
 {
-    internal class AsyncLock
+    internal class AsyncLock : IAsyncDisposable
```

Comment:
It feels like this deserves an extensive set of tests, and I don't see any in this PR.  What you have here looks strictly more correct than what you previously had, though I'd feel better about it if there were tests validating all the scenarios required by its internal usage.

Reply:
The scenario required by it's internal usage is reliable sessions, and we have pretty extensive tests around that functionality going so far as to intercept the underlying HTTP requests to mess around with the timing and/or delivery of messages to trigger all the behaviors that ReliableSessions provides. The original implementation worked from a lock standpoint, but leaked memory.